File: 2023/day08/solution.py
#!/usr/bin/env python3
import sys
from functools import cache
import re
from math import lcm

sys.setrecursionlimit(1_000_000)
DEBUG = 0
FILENAME = 'input.txt'
PART1 = PART2 = True
D4 = [(-1, 0), (0, 1), (1, 0), (0, -1)]

for arg in sys.argv[1:]:
  if arg.startswith('-e'): FILENAME = f'''example{arg[2:] if len(arg)>2 else ''}.txt'''
  elif arg.startswith('-d'): DEBUG = int(arg[2:]) if len(arg) > 2 else 1
  else:
    match arg:
      case '-i':
        FILENAME = 'input.txt'
      case '-p1':
        PART1, PART2 = True, False
      case '-p2':
        PART1, PART2 = False, True
      case '-p0':
        PART1, PART2 = False, False
      case _:
        raise Exception(f'unknown {arg=}')

with open(FILENAME, 'r') as f:
  lines = f.read().splitlines()
directions = lines[0]
regex = re.compile(r'([A-Z]{3}) = \(([A-Z]{3}), ([A-Z]{3})\)', re.MULTILINE)
nodes = {match.group(1): (match.group(2), match.group(3)) for match in re.finditer(regex, ''.join(lines[2:]))}


def part1():
  node = 'AAA'
  steps = 0
  while node != 'ZZZ':
    direction = directions[steps % len(directions)]
    node = nodes[node][0 if direction == 'L' else 1]
    steps += 1
    if DEBUG > 1: print(f'{node=} {direction=} {steps=}')

  print(f'part 1: {steps}')


def part2():
  # current_nodes = [x for x in nodes.keys() if x.endswith('A')]
  # steps = 0
  # while any(not x.endswith('Z') for x in current_nodes):
  #   direction = directions[steps % len(directions)]
  #   current_nodes = [nodes[x][0 if direction == 'L' else 1] for x in current_nodes]
  #   steps += 1
  #   if DEBUG > 1: print(f'{current_nodes=} {direction=} {steps=}')

  # find the length of each loop. steps is lowest common multiple
  loop_lengths = []
  for node in [x for x in nodes if x.endswith('A')]:
    steps = 0
    while not node.endswith('Z'):
      direction = directions[steps % len(directions)]
      node = nodes[node][0 if direction == 'L' else 1]
      steps += 1
      if DEBUG > 1: print(f'{node=} {direction=} {steps=}')
    loop_lengths.append(steps)
  if DEBUG > 0: print(f'{loop_lengths=}')
  steps = lcm(*loop_lengths)
  print(f'part 2: {steps}')


if PART1: part1()
if PART2: part2()

File: 2023/day08/solution.ts
#!/usr/bin/env -S deno --allow-read

const DEBUG = Deno.args.reduce((acc, item) => (item.startsWith('-d') ? Number(item.slice(2) || '1') : acc), 0);
const FILENAME = Deno.args.reduce(
  (acc, item) => (item == '-i' ? 'input.txt' : item.startsWith('-e') ? `example${item.slice(2)}.txt` : acc),
  'example.txt'
);
const [PART1, PART2] = Deno.args.reduce(
  (acc, item) => (item == '-p0' ? [false, false] : item == '-p1' ? [true, false] : item == '-p2' ? [false, true] : acc),
  [true, true]
);

console.log({ FILENAME, DEBUG, PART1, PART2 });

// deno-lint-ignore no-explicit-any
const debug = (level: number, ...data: any[]) => {
  if (DEBUG >= level) console.debug(...data);
};

const input = await Deno.readTextFile(FILENAME);

const parseInput = () => {
  const sections = input.split('\n\n');
  const directions = sections[0].split('').map((c) => (c == 'L' ? 0 : 1));
  const network = new Map(
    sections[1]
      .split('\n')
      .filter((line) => line.trim())
      .map((line) => {
        const tokens = line.split(/[\s=(),]+/);
        return [tokens[0], [tokens[1], tokens[2]]];
      })
  );
  return { directions, network };
};

const part1 = () => {
  const { directions, network } = parseInput();
  debug(1, { network, directions });
  // ugh let
  let node = 'AAA',
    step = 0;
  while (node != 'ZZZ') {
    const direction = directions[step % directions.length];
    debug(1, { node, step, direction });
    node = network.get(node)![direction];
    ++step;
  }
  console.log('part 1:', step);
};

const gcd = (left: number, right: number) => {
  while (right != 0) {
    [left, right] = [right, left % right];
  }
  return left;
};

const lcm = (values: number[]) => values.reduce((acc, item) => (acc * item) / gcd(acc, item), 1);

const part2 = () => {
  const { directions, network } = parseInput();
  const startNodes = network.keys().filter((n) => n.endsWith('A'));
  const loopEnds = new Map<string, number>();
  const loopStarts = new Map<string, number>();
  for (const startNode of startNodes) {
    let node = startNode;
    let step = 0;
    // paths don't loop back to A, which is something i did not consider in the python solve. i suppose the input must intentionally have loop starts which are multiples of the loop lengths
    const history = new Map<string, number>();
    while (1) {
      history.set(node, step);
      const direction = directions[step % directions.length];
      debug(3, { node, step, direction });
      node = network.get(node)![direction];
      ++step; // increment before store
      if (loopEnds.has(startNode) && history.has(node)) {
        debug(2, { msg: 'loop', node, step });
        loopStarts.set(startNode, history.get(node)!);
        break;
      }
      if (node.endsWith('Z')) {
        debug(2, { msg: 'end', node, step });
        if (!loopEnds.has(startNode)) loopEnds.set(startNode, step);
      }
    }
    debug(1, { startNode, loopStart: loopStarts.get(startNode), loopEnd: loopEnds.get(startNode) });
  }
  const cyclesAlign = lcm(loopEnds.values().toArray());
  console.log('part 2:', cyclesAlign);
};

if (PART1) part1();
if (PART2) part2();
